include/libtorrent/utp_stream.hpp

```diff
@@ -420,11 +421,35 @@ struct TORRENT_EXTRA_EXPORT utp_stream
 	}
 
 	template <class Const_Buffers>
-	std::size_t write_some(Const_Buffers const& /* buffers */, error_code& /* ec */)
+	std::size_t write_some(Const_Buffers const& buffers, error_code& ec)
 	{
-		TORRENT_ASSERT(false && "not implemented!");
-		// TODO: implement blocking write. Low priority since it's not used (yet)
-		return 0;
+		TORRENT_ASSERT(!m_write_handler);
+		if (m_impl == nullptr)
+		{
+			ec = boost::asio::error::not_connected;
+			return 0;
+		}
+
+#if TORRENT_USE_ASSERTS
+		size_t buf_size = 0;
+#endif
+
+		for (auto i = buffer_sequence_begin(buffers)
+			, end(buffer_sequence_end(buffers)); i != end; ++i)
+		{
+			add_write_buffer(i->data(), i->size());
+#if TORRENT_USE_ASSERTS
+			buf_size += i->size();
+#endif
+		}
+		std::size_t ret = write_some(true);
+		TORRENT_ASSERT(ret <= buf_size);
+		if(ret == 0)
+		{
+			ec = boost::asio::error::would_block;
```

Comment:
this isn't quite right. if `write_some()` returns 0, it just means it couldn't send any packets *right now*. However, since you called `add_write_buffer()` already, the socket is remembering those pointers and will read from them later, when it *can* send more packets on the socket.

The normal behavior of a blocking write is that the socket/kernel will copy some of the data into its own send buffer, which later will be sent. If that send buffer is full, the call fails with `EWOULDBLOCK`. There isn't a send buffer in the uTP sockets though. Since they only support asynchronous writes, all they support is to remember where the user's buffers are, and read directly from them.

Have you seen this work? does the ssl unit test pass with the GnuTLS backend?

I can't think of a simple way of supporting this. I think we'd either need a send buffer, or have this never buffer anything but just consume any data that can be sent as packets *right now* (which most likely will be 0 most of the time)

Reply:
You are right, the `write_some` I implemented only supports non-blocking immediate writes due to the limitations you mentioned (similar to socket behavior with the non-blocking option set to true), which is OK as long as you don't mix it with asynchronous writes of course. I'll comment about that.
Luckily, it's only what I needed for the GnuTLS wrapper. It allows to make it much simpler as relying on async writes would require buffering data on the wrapper level.